File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        unordered_map<char, int> charMap;
        int l = 0, max_len = 0;
        
        for (int r = 0; r < s.length(); r++) {
            if (charMap.find(s[r]) != charMap.end() && charMap[s[r]] >= l) {
                l = charMap[s[r]] + 1;
            }
            charMap[s[r]] = r;
            max_len = max(max_len, r - l + 1);
        }
        return max_len;
    }
};

File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        HashMap<Character, Integer> charMap = new HashMap<>();
        int l = 0, max_len = 0;
        
        for (int r = 0; r < s.length(); r++) {
            char curr = s.charAt(r);
            if (charMap.containsKey(curr) && charMap.get(curr) >= l) {
                l = charMap.get(curr) + 1;
            }
            charMap.put(curr, r);
            max_len = Math.max(max_len, r - l + 1);
        }
        return max_len;
    }
}
